inflate.c

```diff
@@ -946,8 +985,15 @@ int32_t Z_EXPORT PREFIX(inflate)(PREFIX3(stream) *strm, int32_t flush) {
                 out -= left;
                 strm->total_out += out;
                 state->total += out;
-                if (INFLATE_NEED_CHECKSUM(strm) && (state->wrap & 4) && out)
-                    strm->adler = state->check = UPDATE(state->check, put - out, out);
+
+                if (INFLATE_NEED_CHECKSUM(strm) && out && state->wrap & 4) {
+                    if (state->flags) {
+                        functable.crc32_fold(&state->crc_fold, put - out, out, 0);
```

Comment:
Fuzzer error on this line..
```
==31==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7ff450d9e800 at pc 0x0000005a4de1 bp 0x7ffcd5555590 sp 0x7ffcd5555588
READ of size 16 at 0x7ff450d9e800 thread T0
SCARINESS: 26 (multi-byte-read-heap-buffer-overflow)
    #0 0x5a4de0 in crc32_fold_load /src/zlib-ng/arch/x86/crc32_fold_pclmulqdq.c:235:14
    #1 0x5a4de0 in crc32_fold_pclmulqdq /src/zlib-ng/arch/x86/crc32_fold_pclmulqdq.c:447:9
    #2 0x58790a in zng_inflate /src/zlib-ng/inflate.c:991:25
    #3 0x5657e8 in gz_decomp /src/zlib-ng/gzread.c:180:15
    #4 0x563d05 in gz_fetch /src/zlib-ng/gzread.c:233:17
    #5 0x561f3d in gz_read /src/zlib-ng/gzread.c:312:17
    #6 0x5618f0 in zng_gzread /src/zlib-ng/gzread.c:368:21
    #7 0x55dd5f in gz_uncompress /src/zlib-ng/test/fuzz/fuzzer_minigzip.c:152:15
    #8 0x55e39a in file_uncompress /src/zlib-ng/test/fuzz/fuzzer_minigzip.c:235:5
```

Reply:
Perhaps it can't handle the case when out < 16?

Reply:
Well what's strange about that though is that it's failing on the load of the "fold" buffer, which should always be a fixed size with enough bytes.

Reply:
What's even weirder is that that buffer is way larger than it needs to be, the partial fold that gets stored isn't _actually_ necessary.

Reply:
It may have to finalize the previous fold and then use `functable.crc32` and reset the fold again??

Reply:
Hmm, I'd like to understand better why that gunzip macro existed in the first place.

Reply:
So you may be onto something given that the libpng issue is also in the CRC function.  Where do you think it needs to reset?

Reply:
Hmm, I do kind of wonder if chunkcopy_safe's bug could be rearing it's ugly head, here.  Then again, I'm having trouble producing exactly the same issue locally.

Reply:
How do I make cmake fetch the fuzzer content?  I feel like I've done this successfully once but don't remember how I did it.

Reply:
Hmm, so the place where it's wrong is suspect.  That's basically where it's trying to consume the end of the buffer.  I more or less tried to mimic what it was doing on the head of Develop right now.  I think the math on out may be part of the issue here.

Reply:
Although...this is doing the same thing:
https://github.com/nmoinvaz/zlib-ng/blob/improvements/revert-inf-reorg-v2/inflate.c#L949

Reply:
I'm a bit stumped.  The only way that this can really happen is if something manages to overwrite the pointer inside the crc_fold struct somehow.  Perhaps asan is missing that somehow.

I do know that chunkcopy_safe is _supposed_ to do a self modifying copy loop with an overlap and our current implementation does not emulate that properly.  I suppose I could temporarily make it do the dumb thing just to see if it passes CI with the fuzzer.

Reply:
Hah so....guess what variable just so happens to occur directly before crc_fold...
The inflate window.  The very window those copy functions deal with.  I really do think that this might be chunkcopy_safe manifesting itself.  If it is, we really need to merge #1173.

Reply:
That's what I would do to make sure it is not that.

Reply:
To reproduce fuzzer bug:
1. Download https://storage.googleapis.com/zlib-ng-backup.clusterfuzz-external.appspot.com/corpus/libFuzzer/zlib-ng_fuzzer_minigzip/public.zip
2. Extract it somewhere
3. Use fuzzer_minigzip and pass it bbaa0e485c93c8683c6614c9049b1bb103987909

Reply:
Hmm, yeah my version merged with the fixed chunkcopy didn't fix it.  What's also interesting is the failed CRCs on the libpng tests - the libpng only ever calls the CRC checksum on the compressed stuff, so something is awry even before we begin inflating?

Reply:
Ok so this is both good news and bad news.  Your branch `improvements/revert-inf-reorg-v2` suffers the same failures in libpng.  So...something's not quite right here, not sure what though.

I did notice that updatewindow is called with the inflate dictionary.  Does this produce unintended side effects?

Reply:
There's a distinct possibility that the issue being caught in asan is one that is entirely separate from the libpng issue.  It could very well be that the fuzzer issue goes away with #1173 merged in.

Reply:
```
StandaloneFuzzTargetMain: running 1 inputs
Running: ./fuzzer_minigzip fuzzershit/bbaa0e485c93c8683c6614c9049b1bb103987909
=================================================================
==7509==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7fcdffd75800 at pc 0x7fce02fef6e6 bp 0x7ffdad08c270 sp 0x7ffdad08c260
READ of size 16 at 0x7fcdffd75800 thread T0
    #0 0x7fce02fef6e5 in crc32_fold_pclmulqdq (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x3e6e5)
    #1 0x7fce02fd8ebe in inflate (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x27ebe)
    #2 0x7fce02ff1fa9 in gz_decomp (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x40fa9)
    #3 0x7fce02ff2e05 in gz_fetch (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x41e05)
    #4 0x7fce02ff33d4 in gz_read (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x423d4)
    #5 0x7fce02ff363a in gzread (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x4263a)
    #6 0x55cfc6ea0a63 in gz_uncompress (/scratchdir/adam/zlib-ng-inline_crc_copy/build3/fuzzer_minigzip+0x2a63)
    #7 0x55cfc6ea0f4b in file_uncompress (/scratchdir/adam/zlib-ng-inline_crc_copy/build3/fuzzer_minigzip+0x2f4b)
    #8 0x55cfc6ea13e4 in LLVMFuzzerTestOneInput (/scratchdir/adam/zlib-ng-inline_crc_copy/build3/fuzzer_minigzip+0x33e4)
    #9 0x55cfc6ea0475 in main (/scratchdir/adam/zlib-ng-inline_crc_copy/build3/fuzzer_minigzip+0x2475)
    #10 0x7fce02de62f9 in __libc_start_call_main (/lib64/libc.so.6+0x292f9)
    #11 0x7fce02de63a7 in __libc_start_main_alias_1 (/lib64/libc.so.6+0x293a7)
    #12 0x55cfc6ea05d0 in _start (/scratchdir/adam/zlib-ng-inline_crc_copy/build3/fuzzer_minigzip+0x25d0)

0x7fcdffd75800 is located 0 bytes to the right of 262144-byte region [0x7fcdffd35800,0x7fcdffd75800)
allocated by thread T0 here:
    #0 0x7fce030da98a in __interceptor_posix_memalign (/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.1/libasan.so.6+0xd398a)
    #1 0x7fce02ff261d in gz_look (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x4161d)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/adam/scratch/zlib-ng-inline_crc_copy/build3/libz.so.1+0x3e6e5) in crc32_fold_pclmulqdq
Shadow bytes around the buggy address:
  0x0ffa3ffa6ab0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ffa3ffa6ac0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ffa3ffa6ad0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ffa3ffa6ae0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ffa3ffa6af0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0ffa3ffa6b00:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0ffa3ffa6b10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0ffa3ffa6b20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0ffa3ffa6b30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0ffa3ffa6b40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0ffa3ffa6b50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
```

Reply:
Ok, so looking a little bit more carefully, it looks as though we're just running off the edge of the input.  This makes a great deal more sense than what I thought was happening.  Line 447 is a load from src.

Reply:
So that load is occurring in the loop for if len >= 64, so I suspect that something is coming in as a negative sign for length maybe, and that's translating to an artificially high len.

Reply:
Yes, here's len somehow:
```(gdb) p len
$9 = 18446744073709280703
```

Reply:
Hmm, looks like something might somehow be smashing the stack in that function:
```
crc32_fold_pclmulqdq (crc=0x555555579e10, src=0x5555555fbdad "\271\034p\310\375\fK%Llw\240\307\340{", <incomplete sequence \365>, len=18, init_crc=0) at /home/adam/scratch/zlib-ng-inline_crc_copy/arch/x86/crc32_fold_pclmulqdq.c:396
396	Z_INTERNAL void crc32_fold_pclmulqdq(crc32_fold *crc, const uint8_t *src, size_t len, uint32_t init_crc) {
(gdb) p len
$20 = 18
(gdb) n
400	    __m128i xmm_initial = _mm_cvtsi32_si128(init_crc);
(gdb) 
401	    int32_t first = init_crc != 0;
(gdb) 
407	    assert(len >= 31 || first == 0);
(gdb) 
408	    crc32_fold_load((__m128i *)crc->fold, &xmm_crc0, &xmm_crc1, &xmm_crc2, &xmm_crc3);
(gdb) 
410	    algn_diff = ((uintptr_t)16 - ((uintptr_t)src & 0xF)) & 0xF;
(gdb) 
411	    if (algn_diff) {
(gdb) 
412	        if (algn_diff >= 4) {
(gdb) 
421	            xmm_t0 = _mm_loadu_si128((__m128i*)src);
(gdb) 
422	            xmm_crc_part = _mm_loadu_si128((__m128i*)src + 1);
(gdb) p len
$21 = 18
(gdb) watch len
Hardware watchpoint 3: len
(gdb) c
Continuing.

Hardware watchpoint 3: len

Old value = 18
New value = 18446744073709551615
crc32_fold_pclmulqdq (crc=0x555555579e10, src=0x5555555fbdc0 "", len=18446744073709551615, init_crc=0) at /home/adam/scratch/zlib-ng-inline_crc_copy/arch/x86/crc32_fold_pclmulqdq.c:433
433	    xmm_crc_part = _mm_setzero_si128();
```

Reply:
Yeah... I see what's happening.  You were pretty much correct that we're not handling things properly for undersized buffers.  I should be able to resolve this with enough careful index math.

Reply:
I wonder if this is at all related to the `infcover` segfault in win32 clangcl build that is present in develop.

Reply:
Got it.

Reply:
Not quite out of the woods yet, the same 5 tests in libpng are failing based on invalid IDAT CRC checksums.  That may be a logical issue on our part, though, independent of this issue.

Reply:
So I think the msan issue was a sanitizer detecting a load into a vector register even though only part of it was used.  However, the libpng tests are very reproducible and happen on your branch that just calls the UPDATE macro in updatewindow, so those are independent of the possible CRC folding issues.

I think something's not quite right with how checksums are called with the updatewindow function.  Clue 1 is that it works _most_ of the time.  I'm wondering if the failure cases have a common code path.

Reply:
Could the issue possibly have to do with the Z_FLUSH parameter?

Reply:
I don't think it would have anything to do with that because it doesn't change any of the state variables really besides the window ones which aren't used after the fact.. at least not that I know of.

And it keeps in tack the pending out and that isn't changed.

I wonder how many times inflate is called with Z_FLUSH and if that has anything to do with it. Normally it is called once. But it can be called multiple and the first few times inflate returns Z_BUF_ERROR.

Reply:
I wonder if you can printf out the crc32 src, dest, and Len values each time and see if the same values are called at the end.

Reply:
I think you are probably heading in the right direction.

Reply:
Definitely not Z_FLUSH - it's 0 through all the libpng tests.  Yeah I'll try printing the CRCs to at least find the cases where they differ between working and not working and see if I can determine a pattern.

Reply:
Ok, so all the adler checksums that actually print during the case CHECK: bit seem to match.  What's interesting is that in the failure cases (vs the working), it seems to terminate printing out checksums way earlier.  I'm printing to stderr so the output is not buffered, it just stops printing out checksums.  You'd think that we'd see at least one mismatch for the uncompressed row that failed but we strangely don't.  It's as if the final checksum that's suppose to run never actually does:

![cksums](https://user-images.githubusercontent.com/3620277/158388149-5c6101d0-9d0f-43af-b3cb-adb29159f605.png)

Reply:
So it has occurred to me that case CHECK could possibly be called if state->wrap is not defined or if out is 0.  So I'm doing the exhaustive print of the adler calculations every update window call and for the working case, every UPDATE call.  It's 732MB of output :).

Reply:
I suspect something is not _quite_ right with how we're invoking the checksum within the updatewindow function.  The CI is failing for a mismatched checksum on minigzip, too.

Reply:
So...if we reduce this to just the failing tests and eliminate the calls when the checksum is 0 bytes, we're correct all the way up to the point where the working case (pr #1178 ) ends seemingly computing waaaay more adler checksums.  There's not a mismatched checksum, though.  It just seems like the version that tries to call the checksum at the same time the updatewindow copy routines occur stops abruptly.

![cksum_failing_tests](https://user-images.githubusercontent.com/3620277/158406171-bddb199e-54f6-4964-9927-74882fd0ac33.png)

After all the missing lines, what's interesting is that the checksums pick up again and are in common, for like 2 or 3 runs.  It's not clear if those are different rows or run from a different part of the extraction (perhaps the dictionary?) or what, but those later 2 checksums on 3 bytes each are correct:

![correct_again](https://user-images.githubusercontent.com/3620277/158406533-12f43334-8c4b-4273-a1b6-7f83666fcecd.png)

Reply:
Haaaah, yep that's it:
```state->mode when UPDATE is invoked: 16206                                                                                                                                             
computing adler on 33 bytes, value = e7710ff1                                                                                                                                         
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 15 bytes, value = 2dee056d                                                                                                                                         
computing adler on 15 bytes, value = 2dee056d                                                                                                                                         
state->mode when UPDATE is invoked: 16200 ```

16206 == CHECK.  I think I found the crux of the logical error here.  Now I guess the question is what's the sensible and most efficient solution?

It would seem that the checksums that spit out after that are the same sequence of bytes, so likely there's a redundant checksum being invoked here?  Here's more of the output:

```
state->mode when UPDATE is invoked: 16200                                                                                                                                             
computing adler on 17 bytes, value = 81a0101                                                                                                                                          
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 20 bytes, value = 4e8d06fe                                                                                                                                         
computing adler on 20 bytes, value = 4e8d06fe                                                                                                                                         
computing adler on 17 bytes, value = 81a0101                                                                                                                                          
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 15 bytes, value = 2dee056d                                                                                                                                         
computing adler on 15 bytes, value = 2dee056d                                                                                                                                         
state->mode when UPDATE is invoked: 16206                                                                                                                                             
computing adler on 33 bytes, value = e7710ff1                                                                                                                                         
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 3 bytes, value = 2720138                                                                                                                                           
computing adler on 15 bytes, value = 2dee056d                                                                                                                                         
computing adler on 15 bytes, value = 2dee056d
```

So it's being run on bigger chunks in this case, but seemingly it needs to be run.  Is there perhaps a not stupid path here where the adler/check variable is just copied over?  Or, is this actually intrinsic to what's being checked?  I would have assumed that the adler checksum is read from the header and compared against the checksummed bytes rather than being run through the UPDATE macro twice in this case.

Reply:
This also kind of makes me wonder if this is partly why minigzip is always faster than minideflate.  It seems like minideflate could be checksumming the inflated bits twice.

I put a ridiculous if statement at the bottom of the inf_leave macro to make sure we're doing adler checksums, we have input to checksum, we're not in raw mode, we need checksums, all of that jazz.  If I do this when state->flags is not true, it fails gzip tests, so this is seemingly something very specific to deflate headers.

Reply:
Before calling crc_fold_final in CHECK case I think it needs to call updatewindow in CHECK case. Could that be it?

Reply:
Hah, I may have found another possible culprit....
It's actually underflow for the number of bytes for the copy.

Reply:
Hmm. Actually my original code had that `UPDATE` in the `CHECK` switch case, so that can't be it.

Reply:
So I got it working.  One of the criteria causing updatewindow to not be called is the issue here.  Some of the time that's when state->mode == CHECK, but not all of the time.  There is the wsize criteria as well but...I dunno, it's weird.  I'll have to mess with this some later tonight.

Reply:
I pushed the working copy, I would love some working theories around where my TODO: comment is in inf_leave, though.

Reply:
Your idea of calling updatewindow within the case CHECK: label is probably on the right track but I wasn't sure at all how to do it properly since the update window does end up invoking a copy.  What's in develop does the analogue of that by invoking the window_flush_output function.

If you have a cleaner way to do this that'd be awesome.  For sure in the near future I'd like to unify the checksum + copy functions to be a single function and maybe write a checksum but no copy wrapper so that updatewindow ends up looking much cleaner.  I guess I could write a surrogate wrapper for that for now until I write the adler + copy stuff.

Reply:
Also how do I reproduce that current issue with the fuzzer?  That hash for the test doesn't seem to be in the minigzip fuzz tests unless I'm reading this wrong.